repos/fdbt-reference-data-service/src/retrievers/csv_retriever/main.py

```diff
@@ -2,31 +2,70 @@
 import os
 import logging
 from io import BytesIO
-from zipfile import ZipFile
-from urllib.request import urlopen
+from zipfile import ZipFile, is_zipfile
+import requests
 
 logger = logging.getLogger()
 logger.setLevel(logging.INFO)
 
 s3 = boto3.resource('s3')
 
-
 def lambda_handler(event, context):
     try:
-        resp = urlopen(os.getenv('DATA_URL'))
-        zipfile = ZipFile(BytesIO(resp.read()))
-        target_file = os.getenv('TARGET_FILE')
+        data_url = os.getenv('DATA_URL')
+        backup_data_url = os.getenv('BACKUP_DATA_URL')
+        content_type = os.getenv('CONTENT_TYPE')
+        bucket_name = os.getenv('BUCKET_NAME')
+
+        if data_url is None:
+            raise Exception("There was no DATA_URL environment variable!")
+
+        if backup_data_url is None:
+            raise Exception("There was no BACKUP_DATA_URL environment variable!")
+
+        if content_type is None:
+            raise Exception("There was no CONTENT_TYPE environment variable!")
+
+        if bucket_name is None:
+            raise Exception("There was no BUCKET_NAME environment variable!")
+
+        try:
+            response = requests.get(data_url, allow_redirects = True, timeout=10)
+            logger.info("Main data url used")
+            
+        except:
+            response = requests.get(backup_data_url, allow_redirects = True, timeout=10)
+            logger.info("Backup data url used")
         
-        for filename in zipfile.namelist():
-            if (target_file and filename == target_file) or not target_file:
-                s3.meta.client.upload_fileobj(
-                    zipfile.open(filename),
-                    os.getenv('BUCKET_NAME'),
-                    filename,
-                    ExtraArgs={
-                        'ContentType': os.getenv('CONTENT_TYPE')
-                    }
-                )
+        file = BytesIO(response.content)
+
+        if (is_zipfile(file)):
+            logger.info("File retrieved is a ZIP file")
+
+            zip_file = ZipFile(file)
+
+            target_file = os.getenv('TARGET_FILE')
+
+            for file_name in zip_file.namelist():
+                if (target_file and file_name == target_file) or not target_file:
+                    s3.meta.client.upload_fileobj(
+                        zip_file.open(file_name),
+                        bucket_name,
+                        file_name,
+                        ExtraArgs={
+                            'ContentType': content_type 
+                        }
+                    )
+        else:
+            logger.info("File retrieved is not a ZIP file")
+
+            target_file = os.getenv('TARGET_FILE')
+
+            if target_file is None:
+                raise Exception("There was no TARGET_FILE environment variable!")
```

Comment:
It looks like target_file may not be set sometimes, as in line 50 above, and we don't raise an exception. Are we sure we want to except if there's no target_file environment variable, in the case that the file isn't a zip file?

Reply:
On line 53, it checks for it.